### PR TITLE
[IMP] website, web_editor, website_sale: hide block in desktop display mode

### PR DESCRIPTION
**Pull request by bso-odoo**
Since [1] it is possible to request that some blocks are not displayed
on mobile devices.
This commit adds a similar option to prevent blocks from being displayed
on desktops (i.e. on non-mobile devices).

This PR also makes the hidden blocks available in the "Invisble Elements" depending on the current display mode.

[1]: https://github.com/odoo/odoo/commit/9463f0f889f9dd8da6077895c125da4998a933c0

task-2900730

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/views/snippets/snippets.xml [hunk 1]
```diff
@@ -588,12 +588,17 @@
...
-                   data-img="/website/static/src/img/snippets_options/mobile_invisible.svg"
-                   data-active-img="/website/static/src/img/snippets_options/mobile_visible.svg"/>
+        <we-button class="o_we_link o_we_device fa fa-fw fa-eye" data-toggle-device-visibility="current" data-no-preview="true"
+                   data-name="device_visibility_opt" data-device="both"/>
+        <we-button class="o_we_link o_we_device" data-toggle-device-visibility="both" data-no-preview="true"
+                   data-name="device_visibility_opt" data-device="no_mobile" title="Make visible on all devices"
+                   data-img="/website/static/src/img/snippets_options/mobile_invisible.svg"/>
+        <we-button class="o_we_link o_we_device" data-toggle-device-visibility="both" data-no-preview="true"
+                   data-name="device_visibility_opt" data-device="no_desktop" title="Make visible on all devices"
+                   data-img="/website/static/src/img/snippets_options/desktop_invisible.svg"/>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:601]: In fact, I am not sure I like the UX of this whole feature:
1. The "eye" icon is really not straightforward.
2. When I am on desktop edition, I feel like it would be easier to say "I don't want this to show up on mobile" from there, rather than having to switch to mobile edition then click that button. Did we consider having a very explicit select with the three option: both, desktop only, mobile only ? "Two clicks instead of one" does not seem to be an argument here: it would be clearer IMO and it is not a feature you want to use for many blocks on your page anyway.
3. I don't understand how it is relevant for those items to show up in the "Invisible elements" section the way they do: if I click on them, the related element becomes visible... but would still be hidden in the current device. So I am not sure it is really relevant to allow to edit it then save and see it disappear. For me there are two use cases for this "invisible elements" section:
    - The standard case: there is an invisible element on the page which can show up depending on some visitor action. We need a button to show/hide it during edit mode to be able to edit it. Example: popups, cookie bar, the future column hover, ...
    - The "I want an element to be visible again" case: there is an element that was made invisible in the page but that the user wants now visible again. We had two cases (I think): when making the header or footer invisible via the related visibility option. I think this PR here should be a new case of this: when clicking on the "invisible element" entry of something hidden on the current device: make it visible again (by calling the snippet option).

@pierre-pde What do you think?

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 2]
```diff
@@ -2163,7 +2163,14 @@ var SnippetsMenu = Widget.extend({
             this.invisibleDOMMap = new Map();
             const $invisibleDOMPanelEl = $(this.invisibleDOMPanelEl);
             $invisibleDOMPanelEl.find('.o_we_invisible_entry').remove();
-            const $invisibleSnippets = globalSelector.all().find('.o_snippet_invisible').addBack('.o_snippet_invisible');
+            let isMobile;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:2166]: You can squash both commits IMO. Like explained in my other comment, I don't think it make sense to have the feature without the possibility of not having to switch device display anyway.

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 3]
```diff
@@ -3851,7 +3858,7 @@ var SnippetsMenu = Widget.extend({
     /**
      * Preview on mobile.
      */
-    _onMobilePreviewClick() {
+    async _onMobilePreviewClick() {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:3861]: An event handle should not be async

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 4]
```diff
@@ -3876,6 +3883,7 @@ var SnippetsMenu = Widget.extend({
         for (const gridItemEl of gridItemEls) {
             gridUtils._reloadLazyImages(gridItemEl);
         }
+        await this._updateInvisibleDOM();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:3886]: What `_updateInvisibleDOM` does seems already mutex-protected. As awaiting this would not impact anything as inside an event handler, I would not await. At best I would add a comment like `// This is async but using the main editor mutex` or something like that.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I will not review the second commit further, I feel like if we agree specs have to be changed following my comment on the first commit, the code will potentially be simplified.

File: addons/website/views/snippets/snippets.xml (hunk 1 above)
Comment (bso-odoo) [reply addons/website/views/snippets/snippets.xml]: For the record, logging the outcome of the discussion and the replacement requirements:
- "Device" option as last option (right after "Visibility" for elements that have that option)
- 2 buttons: mobile & desktop, each device can be hidden separately - mutually exclusive but both togglable "off".
- Click on invisible element -> toggle the visibility option (=> not hidden on any device anymore)

File: addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js [hunk 5]
```diff
@@ -674,6 +674,19 @@ export class WysiwygAdapterComponent extends ComponentAdapter {
...
+     * @param event
+     * @returns {*}
+     * @private
+     */
+    _onServiceContextGet(event) {
+        const readOnly = {
+            isMobile: this.websiteService.context.isMobile,
+        };
+        return event.data.callback(readOnly);
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js:689]:
```suggestion
    /**
     * Retrieves the website service context.
     *
     * @private
     * @param {OdooEvent} ev
     */
    _onServiceContextGet(ev) {
        ev.data.callback({
            isMobile: this.websiteService.context.isMobile,
        });
    }
```

- `@private` first
- Missing type for the param definition
- The handler should not "return" something (detail but also used "retrieves" in the text)
- not using "event" but "ev" instead to avoid problems with the event global variable of Chrome
- Removing temporary variable

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 6]
```diff
@@ -3875,6 +3884,12 @@ var SnippetsMenu = Widget.extend({
         for (const gridItemEl of gridItemEls) {
             gridUtils._reloadLazyImages(gridItemEl);
         }
+        // This is async but using the main editor mutex.
+        for (const invisibleOverrideEl of this.getEditableArea().find('.o_snippet_override_invisible')) {
+            invisibleOverrideEl.classList.remove('o_snippet_override_invisible');
+            invisibleOverrideEl.dataset.invisible = 1;
+        }
+        this._updateInvisibleDOM();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:3892]: Add the comment above `_updateInvisibleDOM`

File: addons/website/static/src/js/editor/snippets.options.js [hunk 7]
```diff
@@ -2871,7 +2902,9 @@ options.registry.ConditionalVisibility = options.Class.extend({
      * @override
      */
     async onTargetHide() {
-        this.$target[0].classList.add('o_conditional_hidden');
+        if (this.$target[0].classList.contains('o_snippet_invisible')) {
+            this.$target[0].classList.add('o_conditional_hidden');
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2907]: Why this change ?

File: addons/website/static/src/js/editor/wysiwyg.js [hunk 8]
```diff
@@ -86,6 +86,28 @@ const WebsiteWysiwyg = Wysiwyg.extend({
...
+                    el.dataset.invisible = '1';
+                }
+            }
+        }
+        for (const el of this.$editable[0].querySelectorAll('.d-md-none')) {
+            el.classList.add('o_snippet_desktop_invisible');
+            if (!isMobile) {
+                el.dataset.invisible = '1';
+            }
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/wysiwyg.js:109]: The code for the invisible element is normally confined to the snippets.editor.js file, I don't think that should change. If I am not mistaken:
- Adding `o_snippet_XXX_invisible` classes on editor start should not be needed, those should be saved with the DOM?
- Setting theright `invisible` attribute is normally already automatic for all `.o_snippet_XXX_invisible` element?

We may need a migration script though, indeed.

File: addons/website/static/src/scss/website.wysiwyg.scss [hunk 9]
```diff
@@ -33,6 +33,31 @@ html[data-edit_translations="1"] {
...
 
+.o_snippet_override_invisible {
+    display: block !important;
+    opacity: 70%;
+    position: relative;
+
+    &::before {
+        position: absolute;
+        width: -webkit-fill-available;
+        width: -moz-available;
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.wysiwyg.scss:44]: Is not fit-content supported? Although, I am not sure of the need... maybe add some comments?

File: addons/website/static/src/scss/website.wysiwyg.scss [hunk 10]
```diff
@@ -93,29 +118,27 @@ $i: 1;
...
-                width: 15px;
-                fill: $o-we-color-danger;
+                width: 12px;
+                fill: $o-we-sidebar-content-field-clickable-color;
                 margin-bottom: 0;
-                margin-left: -3px;
 
                 &:hover {
-                    fill: darken($o-we-color-danger, 7.5%);
+                    fill: $o-we-sidebar-content-field-pressed-color;
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.wysiwyg.scss:134]: Is not this the default rules? Can't this be removed?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/js/editor/snippets.options.js (hunk 7 above)
Comment (bso-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: This is indeed not needed anymore - it was related to an intermediate development state, sorry.

File: addons/website/static/src/scss/website.wysiwyg.scss (hunk 10 above)
Comment (bso-odoo) [reply addons/website/static/src/scss/website.wysiwyg.scss:135]: The `fill` property defaults to `rgb(0, 0, 0)`. Since we want the icon to be lighter than the background of the button we need to specify `fill` - and to get some visual feedback we need to specify it for `:hover` as well.

File: addons/website/static/src/scss/website.wysiwyg.scss (hunk 9 above)
Comment (bso-odoo) [reply addons/website/static/src/scss/website.wysiwyg.scss:45]: The `fit-content` value could be used when the icon was an FA character.
Now the icon is an SVG mask - and to avoid noise the rendered unwanted (but mandatory) text is reduced to 0px,

File: addons/website/static/src/js/editor/wysiwyg.js (hunk 8 above)
Comment (bso-odoo) [reply addons/website/static/src/js/editor/wysiwyg.js]: Indeed, I misunderstood: I somehow was convinced that those classes and the `invisible` attribute were not saved.

**Review by qsm-odoo — APPROVED** [review]
@robodoo r+ :partying_face:

**Comment by qsm-odoo** [comment]
@robodoo r+ ([upgrade branch](https://github.com/odoo/upgrade/pull/3910) was preventing the merge as the PR are not linked together because not for master)

**Comment by qsm-odoo** [comment]
Damn you robodoo -_-
